ouroboros-network/src/Ouroboros/Network/PeerSelection/LedgerPeers.hs

```diff
@@ -0,0 +1,259 @@
+          -> NumberOfPeers
+          -> m (StdGen, [RelayAddress])
+pickPeers inRng _ pools _ | Map.null pools = return (inRng, [])
+pickPeers inRng tracer pools (NumberOfPeers cnt) = go inRng cnt []
+  where
+    go :: StdGen -> Word16 -> [RelayAddress] -> m (StdGen, [RelayAddress])
+    go rng 0 picked = return (rng, picked)
+    go rng n picked =
+        let (r :: Word64, rng') = random rng
+            d = maxBound :: Word64
+            x = fromIntegral r % fromIntegral d in
+        case Map.lookupGE (AccPoolStake x) pools of
+             -- XXX We failed pick a peer. Shouldn't this be an error?
+             Nothing -> go rng' (n - 1) picked 
+             Just (ackStake, (stake, relays)) -> do
+                 let (ix, rng'') = randomR (0, NonEmpty.length relays - 1) rng'
+                     relay = relays NonEmpty.!! ix
+                 traceWith tracer $ PickedPeer relay ackStake stake
+                 go rng'' (n - 1) (relay : picked)
+
+
+-- | Run the LedgerPeers worker thread.
+runLedgerPeers :: forall m.
+                      ( MonadAsync m
+                      , MonadTime m
+                      )
+               => StdGen
+               -> Tracer m TraceLedgerPeers
+               -> UseLedgerAfter
+               -> LedgerPeersConsensusInterface m
+               -> ([DomainAddress] -> m (Map DomainAddress (Set SockAddr)))
+               -> STM m NumberOfPeers
+               -> (Maybe (Set SockAddr, DiffTime) -> STM m ())
+               -> m Void
+runLedgerPeers inRng tracer useLedgerAfter LedgerPeersConsensusInterface{..} doResolve
+               getReq putRsp = do
+    go inRng (Time 0) Map.empty
+  where
+    go :: StdGen -> Time -> Map AccPoolStake (PoolStake, NonEmpty RelayAddress) -> m Void
+    go rng oldTs peerMap = do
+        let peerListLifeTime = if Map.null peerMap then 30
+                                                   else 1847 -- Close to but not exactly 30min.
+
+        traceWith tracer WaitingOnRequest
+        numRequested <- atomically getReq
+        traceWith tracer $ RequestForPeers numRequested
+        !now <- getMonotonicTime
+        let age = diffTime now oldTs
+        (peerMap', ts) <- if age > peerListLifeTime
+                             then do
+                                 peers_m <- case useLedgerAfter of
+                                               DontUseLedger -> return Nothing
+                                               UseLedgerAfter slot -> atomically $ lpGetPeers slot
+                                 case peers_m of
+                                      Nothing -> do
+                                          traceWith tracer $ FetchingNewLedgerState 0
+                                          return (Map.empty, now)
+                                      Just peers -> do
+                                          traceWith tracer $ FetchingNewLedgerState $ length peers
+                                          return (accPoolStake peers, now)
```

Comment:
We can float one case into the other.  Even though `GHC` will do that it will be easier to read.
```suggestion
                                 case useLedgerAfter of
                                   DontUseLedger -> 
                                     traceWith tracer $ FetchingNewLedgerState 0
                                     return (Map.empty, now)
                                   UseLedgerAfter slot -> 
                                     peers <- atomically $ lpGetPeers slot
                                     traceWith tracer $ FetchingNewLedgerState $ length peers
                                     return (accPoolStake peers, now)

```